tests/backup_restore_test.go

```diff
@@ -109,131 +109,169 @@ func TestBackupRestoreBasic(t *testing.T) {
+	// common PITR routines
+	initDB := func(t *testing.T) (*sql.DB, func()) {
+		a := require.New(t)
+
+		_, stopFn := mysql.SetupTestInstance(t, port)
+
+		db, err := sql.Open("mysql", fmt.Sprintf("%s@tcp(%s:%d)/mysql", username, localhost, port))
+		a.NoError(err)
+
+		_, err = db.Exec(fmt.Sprintf("CREATE DATABASE %s", database))
+		a.NoError(err)
+
+		_, err = db.Exec(fmt.Sprintf("USE %s", database))
+		a.NoError(err)
+
+		_, err = db.Exec(`
+		CREATE TABLE tbl0 (
+			id INT,
+			PRIMARY KEY (id),
+			CHECK (id >= 0)
+		);
+		`)
+		a.NoError(err)
+		_, err = db.Exec(`
+		CREATE TABLE tbl1 (
+			id INT,
+			pid INT,
+			PRIMARY KEY (id),
+			UNIQUE INDEX (pid),
+			CONSTRAINT FOREIGN KEY (pid) REFERENCES tbl0(id) ON DELETE NO ACTION
+		);
+		`)
+		a.NoError(err)
+
+		return db, stopFn
+	}
+
+	insertRangeData := func(t *testing.T, db *sql.DB, begin, end int) {
+		a := require.New(t)
+
 		tx, err := db.Begin()
 		a.NoError(err)
 		defer tx.Rollback()
 
 		for i := begin; i < end; i++ {
-			_, err := tx.Exec(fmt.Sprintf("INSERT INTO t0 VALUES (%d)", i))
+			_, err := tx.Exec(fmt.Sprintf("INSERT INTO tbl0 VALUES (%d)", i))
 			a.NoError(err)
-			_, err = tx.Exec(fmt.Sprintf("INSERT INTO t1 VALUES (%d, %d)", i, i))
+			_, err = tx.Exec(fmt.Sprintf("INSERT INTO tbl1 VALUES (%d, %d)", i, i))
 			a.NoError(err)
 		}
 
 		err = tx.Commit()
 		a.NoError(err)
 	}
 
-	_, stop := mysql.SetupTestInstance(t, port)
-	defer stop()
-
-	db, err := sql.Open("mysql", fmt.Sprintf("%s@tcp(%s:%d)/mysql", username, localhost, port))
-	a.NoError(err)
-	defer db.Close()
-
-	_, err = db.Exec(fmt.Sprintf("CREATE DATABASE %s", database))
-	a.NoError(err)
-
-	_, err = db.Exec(fmt.Sprintf("USE %s", database))
-	a.NoError(err)
+	validateTbl0 := func(db *sql.DB) {
+		rows, err := db.Query("SELECT * FROM tbl0")
+		a.NoError(err)
+		i := 0
+		for rows.Next() {
+			var col int
+			a.NoError(rows.Scan(&col))
+			a.Equal(i, col)
+			i++
+		}
+		a.NoError(rows.Err())
+		// TODO(dragonly): change to numRowsTime1 when RestoreIncremental is implemented
+		a.Equal(numRowsTime0, i)
+	}
 
-	_, err = db.Exec(`
-	CREATE TABLE t0 (
-		id INT,
-		PRIMARY KEY (id),
-		CHECK (id >= 0)
-	);
-	`)
-	a.NoError(err)
-	_, err = db.Exec(`
-	CREATE TABLE t1 (
-		id INT,
-		pid INT,
-		PRIMARY KEY (id),
-		UNIQUE INDEX (pid),
-		CONSTRAINT FOREIGN KEY (pid) REFERENCES t0(id) ON DELETE NO ACTION
-	);
-	`)
-	a.NoError(err)
+	validateTbl1 := func(db *sql.DB) {
+		rows, err := db.Query("SELECT * FROM tbl1")
+		a.NoError(err)
+		i := 0
+		for rows.Next() {
+			var col1, col2 int
+			a.NoError(rows.Scan(&col1, &col2))
+			a.Equal(i, col1)
+			a.Equal(i, col2)
+			i++
+		}
+		a.NoError(rows.Err())
+		// TODO(dragonly): change to numRowsTime1 when RestoreIncremental is implemented
+		a.Equal(numRowsTime0, i)
+	}
 
-	// insert data to make time point t0
-	insertData(db, a, 0, 10)
+	// test cases
+	t.Run("Buggy Application", func(t *testing.T) {
```

Comment:
If we don't plan to use sub-tests, let's not do a t.Run() inside a test.

Reply:
Yes, other PITR CUJ test cases would share the same database schemas. I want to manage them as sub-tests here.